src/persist-client/src/fetch.rs

```diff
@@ -86,17 +87,39 @@ where
             });
         }
 
-        let fetched_part = fetch_leased_part(
-            part,
+        let value = fetch_batch_part_blob(
+            &part.shard_id,
             self.blob.as_ref(),
-            Arc::clone(&self.metrics),
-            &self.metrics.read.batch_fetcher,
+            &self.metrics,
             &self.shard_metrics,
-            None,
-            self.schemas.clone(),
+            &self.metrics.read.batch_fetcher,
+            &part.key,
         )
-        .await;
-        Ok(fetched_part)
+        .await
+        .unwrap_or_else(|blob_key| {
+            // Ideally, readers should never encounter a missing blob. They place a seqno
+            // hold as they consume their snapshot/listen, preventing any blobs they need
+            // from being deleted by garbage collection, and all blob implementations are
+            // linearizable so there should be no possibility of stale reads.
+            //
+            // If we do have a bug and a reader does encounter a missing blob, the state
+            // cannot be recovered, and our best option is to panic and retry the whole
+            // process.
+            panic!("batch fetcher could not fetch batch part: {}", blob_key)
+        });
```

Comment:
This duplicates the comment / panic. Would be nice to share a bit more, but I'm also 100% fine with the duplication if it ends up being a hassle.

Reply:
Yeah, started with that, but felt like the boilerplate of a wrapper just to do what is essentially an `expect` with more details in the message didn't feel worth it. And then the thing that really tipped me over the edge was that we were passing an `Option<LeasedReaderId>` before and then unwrapping the `None` case of that into "batch fetcher", whereas it's cleanly separated into one panic that always has a `LeasedReaderId` and one that is always for "batch fetcher"

Otoh, now that I'm thinking about it again, I realize we could just push the panic up one level into `fetch_batch_part_blob` and remove the Result. We'd have to flip back to the `Option<LeasedReaderId>`, which is fine. I think I have a very small preference for how it is now, but seems like we've missed the release candidate cut for this week, so if you prefer this option then pretty happy to go with that

Reply:
> seems like we've missed the release candidate cut for this week
Update! We have not. I've got a green build on this, so gonna hit merge to sneak it in and do this bit as a followup PR

Reply:
Sure - I'm fine leaving it. Easy enough to do on demand if it becomes a hassle.

Reply:
Whoops, put together #24895 before I saw this